plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,251 @@
+{
+    PGresult *res = NULL;
+
+    if(PQstatus(ctx->conn) == CONNECTION_OK) {
+        while(PQconsumeInput(ctx->conn) == 0) {
+            res = PQgetResult(ctx->conn);
+            if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+                flb_warn("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+            }
+            PQclear(res);
+        }
+    }
+
+    flb_free(ctx->db_hostname);
+    flb_sds_destroy(ctx->db_table);
+    flb_sds_destroy(ctx->timestamp_key);
+
+    PQfinish(ctx->conn);
+
+    free(ctx);
+    ctx = NULL;
+}
+
+static int cb_pgsql_init(struct flb_output_instance *ins,
+                          struct flb_config *config, void *data)
+{
+
+    struct flb_pgsql_config *ctx;
+    size_t str_len;
+    PGresult *res;
+    char *query = NULL;
+    char *temp = NULL;
+
+    /* set default network configuration */
+    flb_output_net_default(FLB_PGSQL_HOST, FLB_PGSQL_PORT, ins);
+
+    ctx = flb_calloc(1, sizeof(struct flb_pgsql_config));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+
+    ctx->db_hostname = flb_strdup(ins->host.name);
+    snprintf(ctx->db_port, sizeof(ctx->db_port), "%d", ins->host.port);
+
+
+    ctx->db_name = flb_output_get_property("database", ins);
+    if(!ctx->db_name) {
+        ctx->db_name = FLB_PGSQL_DBNAME;
+    }
+
+    ctx->db_table = flb_sds_create(flb_output_get_property("table", ins));
+    if(flb_sds_len(ctx->db_table) == 0) {
+        ctx->db_table = flb_sds_create(FLB_PGSQL_TABLE);
+        if(ctx->db_table == NULL) {
+            pgsql_conf_destroy(ctx);
+            return -1;
+        }
+    }
+
+    ctx->db_user = flb_output_get_property("user", ins);
+    if(!ctx->db_user) {
+        flb_warn("[out_pgsql] You didn't supply a valid user to connect, your current unix user will be used");
+    }
+
+    ctx->db_passwd = flb_output_get_property("password", ins);
+    if(!ctx->db_passwd) {
+        flb_warn("[out_pgsql] You didn't supply a password, you should use a password to authenticate against PostgreSQL");
+    }
+
+    ctx->timestamp_key = flb_sds_create(flb_output_get_property("timestamp_key", ins));
+    if(!ctx->timestamp_key) {
+        ctx->timestamp_key = flb_sds_create(FLB_PGSQL_TIMESTAMP_KEY);
+        if(ctx->timestamp_key == NULL) {
+            pgsql_conf_destroy(ctx);
+            return -1;
+        }
+    }
+
+    flb_info("[out_pgsql] host=%s port=%s dbname=%s ...",
+             ctx->db_hostname, ctx->db_port, ctx->db_name);
+
+    ctx->conn = PQsetdbLogin(ctx->db_hostname,
+                             ctx->db_port,
+                             NULL, NULL,
+                             ctx->db_name,
+                             ctx->db_user,
+                             ctx->db_passwd);
+
+    if(PQstatus(ctx->conn) != CONNECTION_OK) {
+        flb_error("[out_pgsql] failed to connect to host=%s with error: %s",
+                  ctx->db_hostname, PQerrorMessage(ctx->conn));
+
+        pgsql_conf_destroy(ctx);
+
+        return -1;
```

Comment:
Is this necessary?

Could the code be written instead to attempt to connect on initialization, but then retry in each flush?

Reply:
Not possible, because of the async tasks if we cannot ensure a full connection from the begging we should not proceed with the process since this may lead to errors since we will not if the database will exists, table and if we can go async.

Check on every flush will require a lot of work that we don't want to do because in the future the connection may need to use SSL and all that will take a lot of time for the flush, but in the future we should look for solutions to keep a pool of connection.

Reply:
@PettitWesley can we consider this resolved?

Reply:
Ya